File: MacApp/Sources/VTubeLinkService/VTubeStudioManager.swift
import Foundation
import VTubeLinkShared

class VTubeStudioManager: ObservableObject {
    @Published var isConnected = false
    @Published var isAuthenticated = false
    
    private var webSocketTask: URLSessionWebSocketTask?
    private let url = URL(string: "ws://127.0.0.1:8001")!
    
    private let pluginName = "VTube-IFacial-Link-Swift"
    private let pluginDeveloper = "xuan25"
    private let apiVersion = "1.0"
    
    // Flag indicating the injection loop is running to avoid duplicate loops
    private var injectionLoopRunning = false
    // Whether reconnection should be attempted on failure
    private var shouldReconnect = false
    
    // Reference to the captured data source (set by the service)
    weak var dataSource: IFacialMocapReceiver?
    
    func connect() {
        shouldReconnect = true
        let session = URLSession(configuration: .default)
        webSocketTask = session.webSocketTask(with: url)
        webSocketTask?.resume()
        
        DispatchQueue.main.async {
            self.isConnected = true
        }
        
        Task {
            await performInit()
        }
    }
    
    func disconnect() {
        shouldReconnect = false
        injectionLoopRunning = false
        webSocketTask?.cancel(with: .goingAway, reason: nil)
        webSocketTask = nil
        DispatchQueue.main.async {
            self.isConnected = false
            self.isAuthenticated = false
        }
    }
    
    private func reconnect() {
        guard shouldReconnect else { return }
        print("Reconnecting in 3 seconds...")
        DispatchQueue.main.asyncAfter(deadline: .now() + 3) { [weak self] in
            guard let self = self, self.shouldReconnect else { return }
            let session = URLSession(configuration: .default)
            self.webSocketTask = session.webSocketTask(with: self.url)
            self.webSocketTask?.resume()
            self.isConnected = true
            Task {
                await self.performInit()
            }
        }
    }
    
    // Simple send-then-receive helper mirroring Python's `await websocket.send(); await websocket.recv()`
    private func sendAndReceive(_ payload: [String: Any]) async throws -> [String: Any] {
        guard let task = webSocketTask else {
            throw NSError(domain: "VTS", code: -1, userInfo: [NSLocalizedDescriptionKey: "Not connected"])
        }
        
        let messageType = payload["messageType"] as? String ?? "unknown"
        
        let jsonData = try JSONSerialization.data(withJSONObject: payload)
        let jsonString = String(data: jsonData, encoding: .utf8)!
        
        // Send
        print("[VTS] Sending: \(messageType)")
        try await task.send(.string(jsonString))
        print("[VTS] Sent. Waiting for response...")
        
        // Receive the response
        let message = try await task.receive()
        print("[VTS] Received response for: \(messageType)")
        
        switch message {
        case .string(let text):
            if let data = text.data(using: .utf8),
               let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
                return json
            }
            throw NSError(domain: "VTS", code: -2, userInfo: [NSLocalizedDescriptionKey: "Invalid JSON"])
        case .data(_):
            throw NSError(domain: "VTS", code: -3, userInfo: [NSLocalizedDescriptionKey: "Unexpected binary data"])
        @unknown default:
            throw NSError(domain: "VTS", code: -4, userInfo: [NSLocalizedDescriptionKey: "Unknown message type"])
        }
    }
    
    /// Mirrors the Python `vtube.init(websocket)` flow exactly: auth, then register params.
    private func performInit() async {
        do {
            let authSuccess = try await authenticateOrRequestToken()
            if authSuccess {
                try await checkAndRegisterParameters()
                print("[VTS] Successfully initialized. Starting injection loop.")
                await injectionLoop()
            } else {
                print("[VTS] Authentication failed. Not starting injection.")
            }
        } catch {
            print("Init failed: \(error)")
            // Clean up the broken connection and try again
            webSocketTask?.cancel(with: .goingAway, reason: nil)
            webSocketTask = nil
            DispatchQueue.main.async {
                self.isConnected = false
                self.isAuthenticated = false
            }
            reconnect()
        }
    }
    
    private func authenticateOrRequestToken() async throws -> Bool {
        let tokenKey = "vts_auth_token"
        let savedToken = UserDefaults.standard.string(forKey: tokenKey)
        
        var success = false
        
        if let token = savedToken, !token.isEmpty {
            print("[VTS] Found saved token, trying to authenticate...")
            success = try await authenticate(token: token)
            if !success {
                print("[VTS] Token invalid, clearing and requesting new one...")
                UserDefaults.standard.removeObject(forKey: tokenKey)
                let newToken = try await requestToken()
                UserDefaults.standard.set(newToken, forKey: tokenKey)
                success = try await authenticate(token: newToken)
            }
        } else {
            print("[VTS] No token found, requesting one...")
            let newToken = try await requestToken()
            UserDefaults.standard.set(newToken, forKey: tokenKey)
            success = try await authenticate(token: newToken)
        }
        
        return success
    }
    
    private func requestToken() async throws -> String {
        let payload: [String: Any] = [
            "apiName": "VTubeStudioPublicAPI",
            "apiVersion": apiVersion,
            "requestID": UUID().uuidString,
            "messageType": "AuthenticationTokenRequest",
            "data": [
                "pluginName": pluginName,
                "pluginDeveloper": pluginDeveloper
            ]
        ]
        
        let response = try await sendAndReceive(payload)
        guard let data = response["data"] as? [String: Any],
              let token = data["authenticationToken"] as? String else {
            throw NSError(domain: "VTS", code: -10, userInfo: [NSLocalizedDescriptionKey: "No token in response"])
        }
        return token
    }
    
    private func authenticate(token: String) async throws -> Bool {
        let payload: [String: Any] = [
            "apiName": "VTubeStudioPublicAPI",
            "apiVersion": apiVersion,
            "requestID": UUID().uuidString,
            "messageType": "AuthenticationRequest",
            "data": [
                "pluginName": pluginName,
                "pluginDeveloper": pluginDeveloper,
                "authenticationToken": token
            ]
        ]
        
        let response = try await sendAndReceive(payload)
        if let data = response["data"] as? [String: Any],
           let authenticated = data["authenticated"] as? Bool {
            DispatchQueue.main.async {
                self.isAuthenticated = authenticated
            }
            return authenticated
        }
        return false
    }
    
    private func checkAndRegisterParameters() async throws {
        // 1. Fetch existing parameters
        let fetchPayload: [String: Any] = [
            "apiName": "VTubeStudioPublicAPI",
            "apiVersion": apiVersion,
            "requestID": UUID().uuidString,
            "messageType": "InputParameterListRequest"
        ]
        
        var existingCustomParams = Set<String>()
        let fetchResponse = try await sendAndReceive(fetchPayload)
        if let data = fetchResponse["data"] as? [String: Any],
           let customParamsArray = data["customParameters"] as? [[String: Any]] {
            for param in customParamsArray {
                if let name = param["name"] as? String {
                    existingCustomParams.insert(name)
                }
            }
        }
        
        // 2. Register missing parameters
        for paramName in Constants.customParams {
            if !existingCustomParams.contains(paramName) {
                let createPayload: [String: Any] = [
                    "apiName": "VTubeStudioPublicAPI",
                    "apiVersion": apiVersion,
                    "requestID": UUID().uuidString,
                    "messageType": "ParameterCreationRequest",
                    "data": [
                        "parameterName": paramName,
                        "explanation": "",
                        "min": 0,
                        "max": 1,
                        "defaultValue": 0
                    ]
                ]
                _ = try await sendAndReceive(createPayload)
            }
        }
        print("Parameters registered successfully.")
    }
    
    /// Mirrors the Python PluginThread.run_loop() exactly:
    /// ```python
    /// while not self.should_terminate:
    ///     ifacial_data = self.capture_data.read_data()
    ///     parameter_values = build_params_dict(ifacial_data)
    ///     pack = await vtube.inject_params(websocket, parameter_values)
    /// ```
    /// Single-threaded: send → wait for response → repeat. No concurrent operations.
    private func injectionLoop() async {
        injectionLoopRunning = true
        
        while injectionLoopRunning && isConnected && isAuthenticated {
            guard let source = dataSource else {
                // No data source yet; wait a bit
                try? await Task.sleep(nanoseconds: 16_000_000) // 16ms
                continue
            }
            
            let capturedData = source.latestData
            let params = DataMapper.buildParamsDict(from: capturedData)
            
            let paramValues: [[String: Any]] = params.map { ["id": $0.id, "value": $0.value as Any] }
            
            let payload: [String: Any] = [
                "apiName": "VTubeStudioPublicAPI",
                "apiVersion": "1.0",
                "requestID": UUID().uuidString,
                "messageType": "InjectParameterDataRequest",
                "data": [
                    "faceFound": true,
                    "parameterValues": paramValues
                ] as [String : Any]
            ]
            
            do {
                _ = try await sendAndReceive(payload)
            } catch {
                print("Injection error: \(error)")
                break
            }
        }
        
        print("Injection loop ended.")
        injectionLoopRunning = false
        
        // If we broke out due to an error and we should still be running, reconnect
        if shouldReconnect {
            webSocketTask?.cancel(with: .goingAway, reason: nil)
            webSocketTask = nil
            DispatchQueue.main.async {
                self.isConnected = false
                self.isAuthenticated = false
            }
            reconnect()
        }
    }
}

File: WinApp/VTubeStudioManager.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Diagnostics;
using System.Linq;
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace VTubeLink
{
    public class VTubeStudioManager : INotifyPropertyChanged
    {
        private bool _isConnected;
        public bool IsConnected
        {
            get => _isConnected;
            private set
            {
                if (_isConnected != value)
                {
                    _isConnected = value;
                    PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(IsConnected)));
                }
            }
        }

        private bool _isAuthenticated;
        public bool IsAuthenticated
        {
            get => _isAuthenticated;
            private set
            {
                if (_isAuthenticated != value)
                {
                    _isAuthenticated = value;
                    PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(IsAuthenticated)));
                }
            }
        }

        public event PropertyChangedEventHandler? PropertyChanged;

        private ClientWebSocket? _webSocket;
        private readonly Uri _url = new("ws://127.0.0.1:8001");
        
        private const string PluginName = "VTube-IFacial-Link-Windows";
        private const string PluginDeveloper = "xuan25";
        private const string ApiVersion = "1.0";

        private bool _injectionLoopRunning;
        private bool _shouldReconnect;
        private CancellationTokenSource? _cancellationTokenSource;

        public IFacialMocapReceiver? DataSource { get; set; }

        public async Task ConnectAsync()
        {
            _shouldReconnect = true;
            _cancellationTokenSource = new CancellationTokenSource();
            
            await ConnectInternalAsync();
        }

        private async Task ConnectInternalAsync()
        {
            try
            {
                _webSocket = new ClientWebSocket();
                await _webSocket.ConnectAsync(_url, _cancellationTokenSource!.Token);
                IsConnected = true;

                await PerformInitAsync();
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"[VTS] Connection failed: {ex.Message}");
                CleanupConnection();
                Reconnect();
            }
        }

        public void Disconnect()
        {
            _shouldReconnect = false;
            _injectionLoopRunning = false;
            _cancellationTokenSource?.Cancel();

            CleanupConnection();
        }

        private void CleanupConnection()
        {
            if (_webSocket != null)
            {
                if (_webSocket.State == WebSocketState.Open)
                {
                    try { _webSocket.CloseAsync(WebSocketCloseStatus.NormalClosure, "Closing", CancellationToken.None).Wait(500); } catch { }
                }
                _webSocket.Dispose();
                _webSocket = null;
            }
            IsConnected = false;
            IsAuthenticated = false;
        }

        private void Reconnect()
        {
            if (!_shouldReconnect) return;

            Debug.WriteLine("[VTS] Reconnecting in 3 seconds...");
            Task.Delay(3000).ContinueWith(_ => 
            {
                if (_shouldReconnect && _cancellationTokenSource != null && !_cancellationTokenSource.IsCancellationRequested)
                {
                    _ = ConnectInternalAsync();
                }
            });
        }

        public event Action<MappingUpdate>? MappingUpdated;

        private async Task<JsonElement> SendAndReceiveAsync(object payload)
        {
            if (_webSocket == null || _webSocket.State != WebSocketState.Open)
                throw new InvalidOperationException("WebSocket is not connected");

            var token = _cancellationTokenSource!.Token;
            var jsonString = JsonSerializer.Serialize(payload);
            var bytes = Encoding.UTF8.GetBytes(jsonString);
            
            await _webSocket.SendAsync(new ArraySegment<byte>(bytes), WebSocketMessageType.Text, true, token);

            var buffer = new byte[8192];
            var responseBytes = new List<byte>();
            
            WebSocketReceiveResult result;
            do
            {
                result = await _webSocket.ReceiveAsync(new ArraySegment<byte>(buffer), token);
                if (result.MessageType == WebSocketMessageType.Close)
                {
                    throw new InvalidOperationException("WebSocket closed by server.");
                }
                responseBytes.AddRange(buffer.Take(result.Count));
            } while (!result.EndOfMessage);

            var responseString = Encoding.UTF8.GetString(responseBytes.ToArray());
            return JsonSerializer.Deserialize<JsonElement>(responseString);
        }

        private async Task PerformInitAsync()
        {
            try
            {
                bool authSuccess = await AuthenticateOrRequestTokenAsync();
                if (authSuccess)
                {
                    await CheckAndRegisterParametersAsync();
                    AppLogger.Log("VTS", "Successfully initialized. Starting injection loop.");
                    _ = InjectionLoopAsync();
                }
                else
                {
                    AppLogger.Log("VTS", "Authentication failed. Not starting injection.");
                }
            }
            catch (Exception ex)
            {
                AppLogger.Log("VTS", $"Init failed: {ex.Message}");
                CleanupConnection();
                Reconnect();
            }
        }

        private async Task<bool> AuthenticateOrRequestTokenAsync()
        {
            string? savedToken = ConfigManager.Instance.VtsToken;
            bool success = false;

            if (!string.IsNullOrEmpty(savedToken))
            {
                success = await AuthenticateAsync(savedToken);
                if (!success)
                {
                    string newToken = await RequestTokenAsync();
                    ConfigManager.Instance.VtsToken = newToken;
                    ConfigManager.Instance.Save();
                    success = await AuthenticateAsync(newToken);
                }
            }
            else
            {
                string newToken = await RequestTokenAsync();
                ConfigManager.Instance.VtsToken = newToken;
                ConfigManager.Instance.Save();
                success = await AuthenticateAsync(newToken);
            }

            return success;
        }

        private async Task<string> RequestTokenAsync()
        {
            var payload = new
            {
                apiName = "VTubeStudioPublicAPI",
                apiVersion = ApiVersion,
                requestID = Guid.NewGuid().ToString(),
                messageType = "AuthenticationTokenRequest",
                data = new
                {
                    pluginName = PluginName,
                    pluginDeveloper = PluginDeveloper
                }
            };

            var response = await SendAndReceiveAsync(payload);
            if (response.TryGetProperty("data", out var data) && data.TryGetProperty("authenticationToken", out var tokenObj))
            {
                return tokenObj.GetString() ?? "";
            }
            throw new Exception("No token in response");
        }

        private async Task<bool> AuthenticateAsync(string token)
        {
            var payload = new
            {
                apiName = "VTubeStudioPublicAPI",
                apiVersion = ApiVersion,
                requestID = Guid.NewGuid().ToString(),
                messageType = "AuthenticationRequest",
                data = new
                {
                    pluginName = PluginName,
                    pluginDeveloper = PluginDeveloper,
                    authenticationToken = token
                }
            };

            var response = await SendAndReceiveAsync(payload);
            if (response.TryGetProperty("data", out var data) && data.TryGetProperty("authenticated", out var authenticatedObj))
            {
                bool authenticated = authenticatedObj.GetBoolean();
                IsAuthenticated = authenticated;
                return authenticated;
            }
            return false;
        }

        private async Task CheckAndRegisterParametersAsync()
        {
            var fetchPayload = new
            {
                apiName = "VTubeStudioPublicAPI",
                apiVersion = ApiVersion,
                requestID = Guid.NewGuid().ToString(),
                messageType = "InputParameterListRequest"
            };

            var existingParams = new HashSet<string>();
            var response = await SendAndReceiveAsync(fetchPayload);
            if (response.TryGetProperty("data", out var data) && data.TryGetProperty("customParameters", out var customParamsArray))
            {
                foreach (var param in customParamsArray.EnumerateArray())
                {
                    if (param.TryGetProperty("name", out var nameObj))
                    {
                        existingParams.Add(nameObj.GetString()!);
                    }
                }
            }

            foreach (var paramName in Constants.CustomParams)
            {
                if (!existingParams.Contains(paramName))
                {
                    var createPayload = new
                    {
                        apiName = "VTubeStudioPublicAPI",
                        apiVersion = ApiVersion,
                        requestID = Guid.NewGuid().ToString(),
                        messageType = "ParameterCreationRequest",
                        data = new
                        {
                            parameterName = paramName,
                            explanation = "",
                            min = 0,
                            max = 1,
                            defaultValue = 0
                        }
                    };
                    await SendAndReceiveAsync(createPayload);
                }
            }
        }

        private DateTime _lastBroadcastTime = DateTime.MinValue;

        private async Task InjectionLoopAsync()
        {
            _injectionLoopRunning = true;
            var token = _cancellationTokenSource!.Token;

            var receiveBuffer = new byte[8192];

            while (_injectionLoopRunning && IsConnected && IsAuthenticated && !token.IsCancellationRequested)
            {
                if (DataSource == null)
                {
                    await Task.Delay(16, token);
                    continue;
                }

                var capturedData = DataSource.LatestData;
                var paramsList = DataMapper.BuildParamsDict(capturedData);

                var payload = new
                {
                    apiName = "VTubeStudioPublicAPI",
                    apiVersion = ApiVersion,
                    requestID = Guid.NewGuid().ToString(),
                    messageType = "InjectParameterDataRequest",
                    data = new
                    {
                        faceFound = true,
                        parameterValues = paramsList.Select(p => new { id = p.Id, value = p.Value }).ToArray()
                    }
                };

                // Broadcast Mapping Updates
                var now = DateTime.Now;
                if ((now - _lastBroadcastTime).TotalMilliseconds >= 100)
                {
                    var update = new MappingUpdate();
                    foreach (var kvp in capturedData.Blendshapes) update.ArkitParams[kvp.Key] = kvp.Value;
                    foreach (var p in paramsList) update.VtsParams[p.Id] = p.Value;
                    MappingUpdated?.Invoke(update);
                    _lastBroadcastTime = now;
                }

                try
                {
                    var jsonString = JsonSerializer.Serialize(payload);
                    var bytes = Encoding.UTF8.GetBytes(jsonString);
                    await _webSocket!.SendAsync(new ArraySegment<byte>(bytes), WebSocketMessageType.Text, true, token);
                    
                    // Consume the exact response message completely so buffer doesn't desync
                    WebSocketReceiveResult result;
                    do
                    {
                        result = await _webSocket.ReceiveAsync(new ArraySegment<byte>(receiveBuffer), token);
                    } while (!result.EndOfMessage);
                    
                    await Task.Delay(16, token); // roughly 60fps
                }
                catch (Exception ex)
                {
                    AppLogger.Log("VTS", $"Injection error: {ex.Message}");
                    break;
                }
            }

            _injectionLoopRunning = false;

            if (_shouldReconnect && !token.IsCancellationRequested)
            {
                CleanupConnection();
                Reconnect();
            }
        }
    }
}
